src/zope/app/container/testing.py

```diff
@@ -1,7 +1,140 @@
+         u"\N{CYRILLIC SMALL LETTER KA}"
+         u"\N{CYRILLIC SMALL LETTER A}3"][
+         u"\N{CYRILLIC SMALL LETTER PE}"
+         u"\N{CYRILLIC SMALL LETTER A}"
+         u"\N{CYRILLIC SMALL LETTER PE}"
+         u"\N{CYRILLIC SMALL LETTER KA}"
+         u"\N{CYRILLIC SMALL LETTER A}3_1"] = Folder()
+
+    return root
+
+from zope.traversing.interfaces import ITraversable
+from zope.container.interfaces import ISimpleReadContainer
+from zope.container.traversal import ContainerTraversable
+def setUpTraversal():
+    from zope.traversing.testing import setUp
+    setUp()
+    zope.component.provideAdapter(ContainerTraversable,
+                                  (ISimpleReadContainer,), ITraversable)
+
+
+class Place(object):
+
+    def __init__(self, path):
+        self.path = path
+
+    def __get__(self, inst, cls=None):
+        if inst is None:
+            return self
+
+        try:
+            # Use __dict__ directly to avoid infinite recursion
+            root = inst.__dict__['rootFolder']
+        except KeyError:
+            root = inst.rootFolder = buildSampleFolderTree()
+
+        return traverse(root, self.path)
+
+
+class PlacefulSetup(PlacelessSetup):
+
+    # Places :)
+    rootFolder  = Place(u'')
+
+    folder1     = Place(u'folder1')
+    folder1_1   = Place(u'folder1/folder1_1')
+    folder1_1_1 = Place(u'folder1/folder1_1/folder1_1_1')
+    folder1_1_2 = Place(u'folder1/folder1_2/folder1_1_2')
+    folder1_2   = Place(u'folder1/folder1_2')
+    folder1_2_1 = Place(u'folder1/folder1_2/folder1_2_1')
+
+    folder2     = Place(u'folder2')
+    folder2_1   = Place(u'folder2/folder2_1')
+    folder2_1_1 = Place(u'folder2/folder2_1/folder2_1_1')
+
+    folder3     = Place(u"\N{CYRILLIC SMALL LETTER PE}"
+                        u"\N{CYRILLIC SMALL LETTER A}"
+                        u"\N{CYRILLIC SMALL LETTER PE}"
+                        u"\N{CYRILLIC SMALL LETTER KA}"
+                        u"\N{CYRILLIC SMALL LETTER A}3")
+    folder3_1   = Place(u"\N{CYRILLIC SMALL LETTER PE}"
+                        u"\N{CYRILLIC SMALL LETTER A}"
+                        u"\N{CYRILLIC SMALL LETTER PE}"
+                        u"\N{CYRILLIC SMALL LETTER KA}"
+                        u"\N{CYRILLIC SMALL LETTER A}3/"
+                        u"\N{CYRILLIC SMALL LETTER PE}"
+                        u"\N{CYRILLIC SMALL LETTER A}"
+                        u"\N{CYRILLIC SMALL LETTER PE}"
+                        u"\N{CYRILLIC SMALL LETTER KA}"
+                        u"\N{CYRILLIC SMALL LETTER A}3_1")
+
+    def setUp(self, folders=False, site=False):
+        setUp()
+        cSetUp()
+        setUpTraversal()
+        if folders or site:
+            return self.buildFolders(site)
+
+        from zope.security.management import newInteraction
+        newInteraction()
+
+    def tearDown(self):
+        tearDown()
+        cTearDown()
+        # clean up folders and placeful site managers and services too?
```

Comment:
That's currently `zope.component.testing.tearDown`. I think it should be `zope.testing.cleanup.tearDown`, which will `endInteraction()`.